pkg/resource/autonaming/strategy.go

```diff
@@ -0,0 +1,154 @@
+) (opts *plugin.AutonamingOptions, deleteBeforeCreate bool) {
+	return &plugin.AutonamingOptions{
+		ProposedName: urn.Name(),
+		Mode:         plugin.AutonamingModeEnforce,
+	}, true
+}
+
+// disabledAutonaming is an autonaming config that disables autonaming altogether.
+type disabledAutonaming struct{}
+
+func (a *disabledAutonaming) GetOptions(urn.URN, []byte) (opts *plugin.AutonamingOptions, deleteBeforeCreate bool) {
+	return &plugin.AutonamingOptions{
+		Mode: plugin.AutonamingModeDisabled,
+	}, true
+}
+
+// patternAutonaming is an autonaming config that uses a pattern to generate a name.
+type patternAutonaming struct {
+	// Pattern is the pattern to use to generate the name.
+	Pattern string
+	// Enforce, if true, will enforce the use of the generated name, as opposed to proposing it.
+	// A proposed name can still be overridden by the provider, while an enforced name cannot.
+	Enforce bool
+}
+
+func (a *patternAutonaming) GetOptions(urn urn.URN, randomSeed []byte,
+) (opts *plugin.AutonamingOptions, deleteBeforeCreate bool) {
+	mode := plugin.AutonamingModePropose
+	if a.Enforce {
+		mode = plugin.AutonamingModeEnforce
+	}
+	proposedName, hasRandom := generateName(a.Pattern, urn, randomSeed)
+	return &plugin.AutonamingOptions{
+		ProposedName:    proposedName,
+		Mode:            mode,
+		WarnIfNoSupport: a.Enforce,
+	}, !hasRandom
+}
+
+// providerAutonaming represents the configuration for a provider
+type providerAutonaming struct {
+	// Default is the default autonaming config for the provider unless overridden by a more specific
+	// resource config.
+	Default autonamingStrategy
+
+	// Resources maps resource types to their specific configurations
+	// Key format: provider:module:type (e.g., "aws:s3/bucket:Bucket")
+	Resources map[string]autonamingStrategy
+}
+
+// globalAutonaming represents the root configuration object for Pulumi autonaming
+type globalAutonaming struct {
+	// Default is the default autonaming config for all the providers unless overridden by a more specific
+	// provider config.
+	Default autonamingStrategy
+
+	// Providers maps provider names to their configurations
+	// Key format: provider name (e.g., "aws")
+	Providers map[string]providerAutonaming
+}
+
+func (o *globalAutonaming) pluginOptionsForResourceType(resourceType tokens.Type) (autonamingStrategy, bool, error) {
+	token := string(resourceType)
+
+	// Parse resource type into provider and type
+	parts := strings.Split(token, ":")
+	if len(parts) != 3 {
+		return nil, false, fmt.Errorf("invalid resource type format: %s", resourceType)
+	}
+	provider := parts[0]
+
+	// Check type-specific config
+	if pConfig, ok := o.Providers[provider]; ok {
+		if rConfig, ok := pConfig.Resources[token]; ok {
+			return rConfig, false, nil
+		}
+		if pConfig.Default != nil {
+			return pConfig.Default, false, nil
+		}
+	}
+	// Fall back to global config
+	if o.Default != nil {
+		return o.Default, true, nil
+	}
+	return &defaultAutonamingConfig, true, nil
```

Comment:
Feels to me like `o.Default` should always be non-nil, no? Perhaps we should `contract.Assert` this, for instance.

Reply:
Or check in the caller below.